homeassistant/components/recorder/migration.py

```diff
@@ -1879,7 +1932,39 @@ class EntityIDMigration(BaseRunTimeMigration):
 
     required_schema_version = STATES_META_SCHEMA_VERSION
     migration_id = "entity_id_migration"
-    task = EntityIDMigrationTask
+    task = CommitBeforeMigrationTask
+    # We have to commit before to make sure there are
+    # no new pending states_meta about to be added to
+    # the db since this happens live
+
+    @staticmethod
+    def migrate_data(instance: Recorder) -> bool:
+        """Migrate some data, returns True if migration is completed."""
+        return migrate_entity_ids(instance)
+
+    def migration_done(self, instance: Recorder, did_migrate: bool) -> None:
+        """Will be called after migrate returns True."""
+        # The migration has finished, now we start the post migration
+        # to remove the old entity_id data from the states table
+        # at this point we can also start using the StatesMeta table
+        # so we set active to True
+        _LOGGER.debug("Activating states_meta manager as all data is migrated")
+        instance.states_meta_manager.active = True
+        if did_migrate:
+            instance.queue_task(EntityIDPostMigrationTask())
+        else:
+            with contextlib.suppress(SQLAlchemyError):
+                # If ix_states_entity_id_last_updated_ts still exists
+                # on the states table it means the entity id migration
+                # finished by the EntityIDPostMigrationTask did not
+                # because they restarted in the middle of it. We need
+                # to pick back up where we left off.
+                if get_index_by_name(
+                    self.session,
```

Comment:
Not sure we're allowed to use the session here, the session is likely no longer valid after we've exited the context manager? https://github.com/home-assistant/core/blob/27f891dbb34bb894a0f6f53a3d141dd8751090ee/homeassistant/components/recorder/core.py#L792 

We should maybe not store the session in the migrator instances?

Reply:
The original thought was the session should only be used for the purposes of the checking the state of migrations, and updating the `migrations_changes` table. As long as there is no actual migration of data happening in the session, I think its fine to keep passing the session with the understanding that all the RunTimeMigrations checks/updates won't use the session to do any actual migration of data.

Reply:
Sure, but isn't the session invalidated as soon as we've exited the context manager in `recorder/core.py`?

Reply:
Yes it should not be used once the context manager has been exited

Reply:
Fixed